---
title: "BMEG 400E: Assignment 2"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Assignment Overview 

The goal of this assignment is to walk you through the process of creating a pipeline. For this, we will be analyzing 
the data from the same research article as last time (*https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7772509/*) 
by Choi, Won-Young, et al. (2020). This research group was looking at how different epigenetic marks changed in an 
induced Pluripotent Stem Cell (iPSC) when it was differentiated towards a Neuronal Progenitor Cell (NPC). 
For this assignment, we will be looking at one histone mark (H3K27me3) and its behavior in iPSCs. 
This data was created by looking at Chromatin Immunoprecipitation sequencing data (ChIP-seq) using an antibody 
against the H3K27me3 to sequence only the DNA that attached to the epigenetic mark. Then, we can use this data to 
find out where H3K27me3 marks are located in the genome by mapping the ChIP-seq reads to the genome. However, to 
make sure we are seeing a true enrichment of a region, we need a control to compare to. This control is called the 
*input*, which is usually essentially the same proceedure as was applied to the ChIP-seq sample, but with the 
immunoprecipitation step skipped. By comparing the input to the ChIP data, we can distinguish between noise and true 
enrichements. For this assignment, we will be finding how H3K27me3 changes when iPSCs undergo differentiation. 
All fastq files are paired-end and can be found under the following path:**/projects/bmeg/A2/**. As always, 
remember not to copy these files to your directory or try to alter them in any way!

For iPSC:

  - Input for iPSC: *input_iPSC_SRA66_subset_1.fastq.gz* and *input_iPSC_SRA66_subset_2.fastq.gz*

  - H3K27me3 for iPSC: *H3K27me3_iPSC_SRA60_subset_1.fastq.gz* and *H3K27me3_iPSC_SRA60_subset_2.fastq.gz*

We will not be using the NPC data in this assignment.
  

**This assignment follows a simplified version of a ChIP-seq analysis prior to the peak callling using the following steps:**
  a. Analyze the reads quality: fastqc (*https://www.bioinformatics.babraham.ac.uk/projects/fastqc/*)
  b. Mapping the reads to the genome: bowtie2 (*http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml*)
  c. Convert sam to bam files: samtools (*http://www.htslib.org/doc/*)
  d. Sort the alignment files by genomic position: sambamba (*https://lomereiter.github.io/sambamba/docs/sambamba-view.html*)
  e. Filter for uniquely mapped reads: sambamba 

## Before we get started
Before we get started we will assume that you have successfully created a Github repository that will host all your 
future assignments. For this assignment make sure to create a new file under the same repository so that you don't need 
to make a new one. You can just make a new subdirectory. At the end, all you should need to do is the standard "commit, 
pull, push" actions when you are done the assignment. If you make a new one, remember to keep it private and add this 
accound as a collaborator.

All the questions that need to be answered with a command or a written response, from now on, will be marked with 
**#?#** and be followed by the grading of the question. 

Remember to read carefully each step and to use the tips only as tips and should not be used as instructions on how 
to do things! You are welcome to use other tools and commands you feel more comfortable with **as long as you explain 
your thought process and show that you reached the same goal**. Also remember to pay attention to the error messages 
and try to understand what is it saying, Google it and try a few things by yourself! Most of the times looking at the 
tools documentation would solve it. As a final note, these tools can be daunting, so don't go into any rabbit holes, 
less is more, you won't need to modify any of the default parameters, just make sure you are getting using the right 
input and output files. Okay, let's start!

**This assignment has 3 main goals:**
  1. To go through *steps a-e* for the **input data**
  2. To create a pipeline that does a-e automatically for the **H3K27me3 data**

## 0. Getting the right tools 
As you might have seen on the overview we will be using lots of new tools today. So before jumping into the analysis, 
let's make sure we install them all. Luckily for us, we are all familiar with how to install things using conda (see your 
assignment 1)!

```{bash, eval=FALSE}
#?# Add sambamba to your conda environment created on assignment 1 - 1pt

## Note: to complete this assignment, it is expected that your conda environment already has installed: fastqc, bowtie2 
and samtools 

(base) elamoureux_bmeg22@orca01:~$ conda activate bmeg591e
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments$ conda install -c bioconda sambamba
...
The following NEW packages will be INSTALLED:

  sambamba           bioconda/linux-64::sambamba-0.6.6-2


Proceed ([y]/n)? y


Downloading and Extracting Packages
sambamba-0.6.6       | 1.1 MB    | #################################################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```


## 1. Analyzing ChIP-seq data (steps a-e)
For this first part of the assignment, we will work **only with the iPSC input data.** 

### a. Quality Controls
Similarly to the last assignment, we need to make sure that the data has a good quality before we can move on to 
further steps. The files that we are working with are paired-end, meaning that there are two reads per each sequence, 
one starting on the 5' (_1) and the other on the 3' (_2). Using the **fastqc** tool that we reviewed last time, do a quick 
quality check of the two files included in your set (_1 and _2). 

```{bash, eval=FALSE}

## NOTE: Remember to use screen activate your conda environment!


#?# Type below the command that you used to perform the fastqc analysis on the paired-end iPSC input files: - 0.5 pt
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ dir
H3K27me3_iPSC_SRA60_subset_1.fastq.gz  input_iPSC_SRA66_subset_1.fastq.gz
H3K27me3_iPSC_SRA60_subset_2.fastq.gz  input_iPSC_SRA66_subset_2.fastq.gz
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ gzip -d input_iPSC_SRA66_subset_1.fastq.gz
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ gzip -d input_iPSC_SRA66_subset_2.fastq.gz
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ dir
H3K27me3_iPSC_SRA60_subset_1.fastq.gz  input_iPSC_SRA66_subset_1.fastq
H3K27me3_iPSC_SRA60_subset_2.fastq.gz  input_iPSC_SRA66_subset_2.fastq
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ fastqc input_iPSC_SRA66_subset_1.fastq
Started analysis of input_iPSC_SRA66_subset_1.fastq
Approx 5% complete for input_iPSC_SRA66_subset_1.fastq
Approx 10% complete for input_iPSC_SRA66_subset_1.fastq
Approx 15% complete for input_iPSC_SRA66_subset_1.fastq
Approx 20% complete for input_iPSC_SRA66_subset_1.fastq
Approx 25% complete for input_iPSC_SRA66_subset_1.fastq
Approx 30% complete for input_iPSC_SRA66_subset_1.fastq
Approx 35% complete for input_iPSC_SRA66_subset_1.fastq
Approx 40% complete for input_iPSC_SRA66_subset_1.fastq
Approx 45% complete for input_iPSC_SRA66_subset_1.fastq
Approx 50% complete for input_iPSC_SRA66_subset_1.fastq
Approx 55% complete for input_iPSC_SRA66_subset_1.fastq
Approx 60% complete for input_iPSC_SRA66_subset_1.fastq
Approx 65% complete for input_iPSC_SRA66_subset_1.fastq
Approx 70% complete for input_iPSC_SRA66_subset_1.fastq
Approx 75% complete for input_iPSC_SRA66_subset_1.fastq
Approx 80% complete for input_iPSC_SRA66_subset_1.fastq
Approx 85% complete for input_iPSC_SRA66_subset_1.fastq
Approx 90% complete for input_iPSC_SRA66_subset_1.fastq
Approx 95% complete for input_iPSC_SRA66_subset_1.fastq
Analysis complete for input_iPSC_SRA66_subset_1.fastq
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ fastqc input_iPSC_SRA66_subset_2.fastq
Started analysis of input_iPSC_SRA66_subset_2.fastq
Approx 5% complete for input_iPSC_SRA66_subset_2.fastq
Approx 10% complete for input_iPSC_SRA66_subset_2.fastq
Approx 15% complete for input_iPSC_SRA66_subset_2.fastq
Approx 20% complete for input_iPSC_SRA66_subset_2.fastq
Approx 25% complete for input_iPSC_SRA66_subset_2.fastq
Approx 30% complete for input_iPSC_SRA66_subset_2.fastq
Approx 35% complete for input_iPSC_SRA66_subset_2.fastq
Approx 40% complete for input_iPSC_SRA66_subset_2.fastq
Approx 45% complete for input_iPSC_SRA66_subset_2.fastq
Approx 50% complete for input_iPSC_SRA66_subset_2.fastq
Approx 55% complete for input_iPSC_SRA66_subset_2.fastq
Approx 60% complete for input_iPSC_SRA66_subset_2.fastq
Approx 65% complete for input_iPSC_SRA66_subset_2.fastq
Approx 70% complete for input_iPSC_SRA66_subset_2.fastq
Approx 75% complete for input_iPSC_SRA66_subset_2.fastq
Approx 80% complete for input_iPSC_SRA66_subset_2.fastq
Approx 85% complete for input_iPSC_SRA66_subset_2.fastq
Approx 90% complete for input_iPSC_SRA66_subset_2.fastq
Approx 95% complete for input_iPSC_SRA66_subset_2.fastq
Analysis complete for input_iPSC_SRA66_subset_2.fastq
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ls
H3K27me3_iPSC_SRA60_subset_1.fastq.gz  input_iPSC_SRA66_subset_1_fastqc.html  input_iPSC_SRA66_subset_2_fastqc.html
H3K27me3_iPSC_SRA60_subset_2.fastq.gz  input_iPSC_SRA66_subset_1_fastqc.zip   input_iPSC_SRA66_subset_2_fastqc.zip
input_iPSC_SRA66_subset_1.fastq        input_iPSC_SRA66_subset_2.fastq

## Copy the html output file to your computer, open it and write below what can you tell about the data quality of these 
## files.
## You can reference fastqc documentation if in doubt.

In Windows Command Prompt:
C:\Users\eriks>set PATH=C:\Users\eriks\bmeg_591e\pscp.exe

C:\Users\eriks\bmeg_591e>pscp
PuTTY Secure Copy client
Release 0.76
Usage: pscp [options] [user@]host:source target
       pscp [options] source [source...] [user@]host:target
       pscp [options] -ls [user@]host:filespec
Options:
  -V        print version information and exit
  -pgpfp    print PGP key fingerprints and exit
  -p        preserve file attributes
  -q        quiet, don't show statistics
  -r        copy directories recursively
  -v        show verbose messages
  -load sessname  Load settings from saved session
  -P port   connect to specified port
  -l user   connect with specified username
  -pw passw login with specified password
  -1 -2     force use of particular SSH protocol version
  -ssh -ssh-connection
            force use of particular SSH protocol variant
  -4 -6     force use of IPv4 or IPv6
  -C        enable compression
  -i key    private key file for user authentication
  -noagent  disable use of Pageant
  -agent    enable use of Pageant
  -no-trivial-auth
            disconnect if SSH authentication succeeds trivially
  -hostkey keyid
            manually specify a host key (may be repeated)
  -batch    disable all interactive prompts
  -no-sanitise-stderr  don't strip control chars from standard error
  -proxycmd command
            use 'command' as local proxy
  -unsafe   allow server-side wildcards (DANGEROUS)
  -sftp     force use of SFTP protocol
  -scp      force use of SCP protocol
  -sshlog file
  -sshrawlog file
            log protocol details to a file
  -logoverwrite
  -logappend
            control what happens when a log file already exists

C:\Users\eriks>pscp elamoureux_bmeg22@orca1.bcgsc.ca:assignments/assignment_2/input_iPSC_SRA66_subset_1_fastqc.html c:\users\eriks\bmeg_591e\input_iPSC_SRA66_subset_1_fastqc.html
elamoureux_bmeg22@orca1.bcgsc.ca's password:
input_iPSC_SRA66_subset_1 | 528 kB | 528.7 kB/s | ETA: 00:00:00 | 100%

C:\Users\eriks>pscp elamoureux_bmeg22@orca1.bcgsc.ca:assignments/assignment_2/input_iPSC_SRA66_subset_2_fastqc.html c:\users\eriks\bmeg_591e\input_iPSC_SRA66_subset_2_fastqc.html
elamoureux_bmeg22@orca1.bcgsc.ca's password:
input_iPSC_SRA66_subset_2 | 532 kB | 532.3 kB/s | ETA: 00:00:00 | 100%

#?# Are there any differences between the files for read 1 and read 2? Describe what you see below: - 1.5 pt
## - Overall, the quality of reads 1 and 2 are very good. For both, the basic statistics, per base sequence quality, 
per sequence quality scores, per base sequence content, per base N content, sequence length distribution, sequence duplication
levels, overrepresented sequences, and adapter content were all good and in the green. The only item that was yellow/warning/
incomplete was the per sequence GC content for both reads. 
## - Despite the overall similarities and high quality, read 2 (subset_2) had a greater reduction of quality in reads 
starting at position 105-109 and continued decreasing until the error bars at position 150-151 dipped into the red section. 


## NOTE: Same as last time, in order to open the html files on your web browser you will need to download the files to 
your computer
## Ubuntu/Linux and Mac users: look at the *scp* command
## Windows users: you can follow these instructions: 
https://stackoverflow.com/questions/6217055/how-can-i-copy-a-file-from-a-remote-server-to-using-putty-in-windows

See Windows command prompt outputs above. 
```

### b. Mapping to the reference genome

**IMPORTANT**: This step can take up to ~30mins to run, please be mindful of the resources in the server (use htop) 

```{bash, eval=FALSE}

#?# Perform a paired-end alignment of the iPSC input fastq sequences to the human genome using bowtie2 - 1.5 pt
## Use the previously created index located in: /projects/bmeg/indexes/hg38/hg38_bowtie2_index 

Run htop to check server resources: Mem - 80.5G/126G, 13 Tasks, 1 thread, 4 running. --> enough resources available 
to run upcoming tasks. 


(bmeg591e) elamoureux_bmeg22@orca01:~$ bowtie2 -x /projects/bmeg/indexes/hg38/hg38_bowtie2_index -1 /projects/bmeg/A1/input_iPSC_SRA66_subset_1.fastq -2 /projects/bmeg/A1/input_iPSC_SRA66_subset_2.fastq -S output.sam
2614910 reads; of these:
  2614910 (100.00%) were paired; of these:
    220030 (8.41%) aligned concordantly 0 times
    2129747 (81.45%) aligned concordantly exactly 1 time
    265133 (10.14%) aligned concordantly >1 times
    ----
    220030 pairs aligned concordantly 0 times; of these:
      169387 (76.98%) aligned discordantly 1 time
    ----
    50643 pairs aligned 0 times concordantly or discordantly; of these:
      101286 mates make up the pairs; of these:
        34076 (33.64%) aligned 0 times
        30311 (29.93%) aligned exactly 1 time
        36899 (36.43%) aligned >1 times
99.35% overall alignment rate

(bmeg591e) elamoureux_bmeg22@orca01:~$ mv output.sam ~/assignments/assignment_2/bowtie2_output.sam

## Tip: look at the bowtie2 --help or the documentation on the tool website 
(http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example)

```

### c. Sam to Bam file convertion 

Next we must convert the sam files made by bowtie2 to bam files, which are much smaller in size and faster to access.

```{bash, eval=FALSE}
#?# Using *samtools view*, convert the sam file containing the result of the alignment of the iPSC input files to the 
reference genome (last step output) to bam format - 1.5 pt
## Don't forget to include the header! (-h flag)

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ samtools view -S -b -h bowtie2_output.sam > bowtie2_output.bam
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ls
bowtie2_output.bam                    
bowtie2_output.sam
```

### d. Bam indexing and sorting 

BAM files are much smaller and faster to access than SAM files, but unless the reads are sorted by chromosome and position, 
we can't do much with it since the reads relevant for one locus would be scattered randomly throughout in whatever order 
they had been aligned. Further, indexing is important to allow programs like IGV to access the BAM file. Indexing is 
essentially adding little bookmarks throughout the file so that a program accessing the BAM knows where to find data 
relevant to whatever locus it is interested in. The `sambamba` tool can be used to sort bam files and create an index 
simultaneously. You may need to install it via anaconda.

```{bash, eval=FALSE}

## Run the following command, changing the step_d_output.bam for the name of the bam output file you got from step c; 
this will print the read ID, chromosome, and genomic position for each read alignment; replace <step_c_output>.bam with 
the name of your bam file from the last step.
samtools view <step_c_output>.bam | cut -f1,3,4 | head -5

#?# Use the documentation format (# at the beginning of the line) to include the output of the command above 
(one per line): - 0.5pt

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ samtools view bowtie2_output.bam | cut -f1,3,4 | head -5
SRR12694366.1.1000000   chr3    46631700
SRR12694366.1.1000000   chr3    46631857
SRR12694366.1.10000029  chr3    77719001
SRR12694366.1.10000029  chr3    77718984
SRR12694366.1.10000035  chr3    40015576

## Using *sambamba sort*, sort the bam file that you created on the last step
## sambamba sort default will sort the file and build an index that will allow us to look at the reads that mapped to a 
specific positions in the genome
#?# Type the command you use below: - 1 pt
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ sambamba sort bowtie2_output.bam
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ls
bowtie2_output.sorted.bam              
bowtie2_output.sorted.bam.bai          
bowtie2_output.bam                     
bowtie2_output.sam

## View the read ID, chromosome, and genomic position for the first 5 reads, as before, but this time for the sorted 
bam file you just made.
samtools view step_d_output.bam | cut -f1,3,4 | head -5

#?# Use the documentation format (# at the beginning of the line) to include the output of the command above 
(one per line): 0.5 pt
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ samtools view bowtie2_output.sorted.bam | cut -f1,3,4 | head -5
SRR12694366.1.26500791  chr3    9988
SRR12694366.1.33624963  chr3    10013
SRR12694366.1.33624963  chr3    10013
SRR12694366.1.33999302  chr3    10033
SRR12694366.1.33999302  chr3    10033

#?# What changed between the two files? Describe what is sambamba sort doing. - 1 pt
## If needed, you can inspect more of each bam file by increasing the -n parameter for `head` in the above 
(just don't bother to include more than 5 lines for each in your submission).

What sambamba sort is doing is sorting the entries from ascending order according to the genomic position (e.g. start 9988). Alternatively, 
samtools view on bowtie2_output.bam has the entries in ascending order according to read ID (e.g. start SRR12694366.1.1000000).

```

### e. Keep only uniquely mapped reads 

Next, we want to create a version of the BAM file that contains only uniquely mapping reads. We will be using these 
data to call peaks and want to know which are different between cell types. The reads that do not map uniquely are 
stochastically assigned to one of the several best possible matches. This could lead to regions that have more reads 
in one cell type vs the other by chance alone. Accordingly, we want to remove these ambiguously mapping reads.

```{bash, eval=FALSE}

#?# Use *sambamba view* to filter the recently sorted bam file (previous step output) to include only uniquely mapped 
reads - 1 pt
## For this we will make use of the *-F* flag to filter the reads that were mapped and are not duplicates, by adding 
the following flag:
## *  -F "[XS] == null and not unmapped and not duplicate"  *
## Important: Remember to add the header (-h flag) and to specify the output file format (-f flag) as bam

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ sambamba view -h -f "bam" -F "[XS] == null and not unmapped and not duplicate" bowtie2_output.sorted.bam > sambamba_output_sorted.bam
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ls
...
sambamba_output_sorted.bam


#?# How many reads were there before filtering for uniquely mapped reads? How many are there now? Include the code to 
answer these questions and the answers to them below.

Before filtering:
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ samtools view bowtie2_output.bam | cut -f1,3,4 | wc
5229820 15689460 194447545

Therefore, there are 5,229,820 reads prior to filtering (lines). 

After filtering:
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ samtools view sambamba_output_sorted.bam | cut -f1,3,4 | wc
4541630 13624890 168884955

Therefore, there are 4,541,630 uniquely mapped reads (lines). 
```

Now that you have created your BAM files and inspected them, now would be a good time to delete the SAM files that are 
leftover from your alignment.

```{bash, eval=FALSE}
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ rm bowtie2_output.sam
```

The next step in the process would be calling peaks, but we need both input and ChIP data for each cell type before we 
can do that and here we have only processed one file. So instead, we will be...

## 2. Implementing a pipeline 

We have gone through all these steps for this first file, the input iPSC file. Now we want to make a pipeline that 
could be used for all four files where we have abstracted away the sample IDs, and instead have code that will work 
on any given sample name.

In this section, you will need to edit files on the command line. There are several ways in which you can do this. The 
simplest text editor is `pico`, with `emacs` being more intermediate, and `vim` being powerful, but difficult to learn. 
You can find tutorials for any of these using google. It would be best for you to learn one of these, rather than 
constantly moving files back and forth to/from your computer and the server.

### a. Make a task list

Start by making a tab delimited file with three columns: sampleID, fastqFile1, fastqFile2
A sample first line is included here:

```{bash, eval=FALSE}

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ printf "sampleID\t fastqFile1\t fastqFile2\n" | column -t > task_list.txt
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ head task_list.txt
sampleID  fastqFile1  fastqFile2

printf "iPSC_input\t input_iPSC_SRA66_subset_1.fastq.gz\t input_iPSC_SRA66_subset_2.fastq.gz\n" | column -t > task_list.txt

nano 
# Enter text editor. ctrl(^) R to read file --> read task_list.txt
# Edit text document to read:
```

```{}

iPSC_input  input_iPSC_SRA66_subset_1.fastq.gz  input_iPSC_SRA66_subset_2.fastq.gz
H3K27me3  H3K27me3_iPSC_SRA60_subset_1.fastq.gz  H3K27me3_iPSC_SRA60_subset_2.fastq.gz


```

```{bash, eval=FALSE}

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ head task_list.txt
iPSC_input  input_iPSC_SRA66_subset_1.fastq.gz  input_iPSC_SRA66_subset_2.fastq.gz
H3K27me3  H3K27me3_iPSC_SRA60_subset_1.fastq.gz  H3K27me3_iPSC_SRA60_subset_2.fastq.gz

This will be the list of tasks you want to accomplish. 
```

### b. Working with a "job scheduler"

Normally we would use a job scheduler to accomplish these tasks, but our server is incapable of using a job scheduler, 
so instead we're going to use this program 
(https://raw.githubusercontent.com/BMEGGenoInfo/Assignments/main/Assignment_2/runTheseJobsSerially.sh) that will run a 
given script for each line in a file. Download it to the server, for instance using `wget`. 

Now run the following command, where you should replace <taskfile> with the name of your file. Note that you may need to 
run `chmod +x runTheseJobsSerially.sh` to give the script file execute permissions.
  
```{bash, eval=FALSE}

(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ./runTheseJobsSerially.sh echo task_list.txt
bash: ./runTheseJobsSerially.sh: Permission denied
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ chmod +x runTheseJobsSerially.sh
(bmeg591e) elamoureux_bmeg22@orca01:~/assignments/assignment_2$ ./runTheseJobsSerially.sh echo task_list.txt
iPSC_input input_iPSC_SRA66_subset_1.fastq.gz input_iPSC_SRA66_subset_2.fastq.gz
H3K27me3 H3K27me3_iPSC_SRA60_subset_1.fastq.gz H3K27me3_iPSC_SRA60_subset_2.fastq.gz

./runTheseJobsSerially.sh echo <taskfile>

#?# what happened? Enter your answer below - 1 pt
## -- It seems that runTheseJobsSerially.sh prints back the contents of the task list. 
```


### c. Your first pipeline

Now we want to make a script that will run all the commands you performed above (in Q1), but for any of the samples. 
Below is a script which you can copy into a new file "fastqToFilteredBam.sh", which you will modify to include all the 
steps needed to go from a fastq to a sorted, filtered (uniquely mapping only) bam file.

```{bash, eval=FALSE}

#!/bin/bash
set -e # this makes the whole script exit on any error.
#fill these variables with the arguments given to this script
sample=$1
fq1=$2
fq2=$3
logDir=MyLogDirectory # this is where all the files to keep track of progress will go.
mkdir -p MyLogDirectory # make the directory where log files will go, if it doesn't exist already echo running pipeline for $sample
if [ ! -e $logDir/$sample.fastqc.done ] #run this code only if $logDir/$sample.fastqc.done is missing
then
        echo Performing fastqc of sample $sample with the following fastqs:
        ls /projects/bmeg/A2/$fq1 /projects/bmeg/A2/$fq2

        #enter commands to run fastqc here
        fastqc $fq1
        fastqc $fq2

        touch $logDir/$sample.fastqc.done #create the file that we were looking for at the beginning of this if statement so that this same code is not run next time
else # $logDir/$sample.fastqc.done was not missing
echo Already performed fastqc of $sample

fi

#here is where you will be including the additional steps to take you from fastq.gz to sorted BAM containing only 
uniquely mapping reads.

```


Now run the following:

```{bash, eval=FALSE}

## Run this with <taskfile> replaced with the name of your task file.
./runTheseJobsSerially.sh ./fastqToFilteredBam.sh task_list.txt

#?# What happened? Enter your answer below - 1 pt
## -- For all the fastq files in task_list.txt, this script executes the fastq function.

#?# Now run that same command again. What happened this time? Why? Answer below.  - 2 pts
## -- When it is run again, the script outputs: "Already performed fastqc of iPSC_input" because iPSC_input.fastqc.done exists in
# MyLogDirectory. 

#?# What can you do to make it run the same way as it did the first time? Answer below with the command(s) that would 
make it run the same way as the first time - 2 pts
## -- cd MyLogDirectory
## rm iPSC_input.fastqc.done # this removes the 'done' file and the script can run again as it did the first time. 

```


### d. Filling in the pipeline

Now fill in the rest of the pipeline with the tasks you had done for iPSC input, but replace the sample IDs with the 
variable `$sample`. Include appropriate documentation and messages. Note that it is easier to separate the different 
parts of a file name with a period (.). For instance, bash will treat the `$sample` in `$sample.sorted.uniqMapping.bam` 
as the variable `$sample`, whereas `$sample_sorted_uniqMapping.bam` will be interpreted as a new varaible 
`$sample_sorted_uniqMapping`.

Note that in the script provided, fastqc is not actually run, so you will have to include the code for that too. Everything 
you needed to do to take the samples from fastq to sorted, uniquely mapping BAM file should be included in this pipeline. 
You should also build in robustness to failure, comments, and messages (e.g. via `echo`) to describe in the purpose of 
the code, and to provide updates while the code is running. You should test this script by running it on the two samples 
(iPSC input, and iPSC ChIP). 


```{bash, eval=FALSE}
## Test the script by running this, with your modified pipeline script in place of fastqToFilteredBam2.sh

./runTheseJobsSerially.sh ./fastqToFilteredBam2.sh task_list.txt

```


When you are done, enter the code for your pipeline in the space provided below - 10 pts

```{bash, eval=FALSE}


#?# REPLACE THIS WITH YOUR SCRIPT

#!/bin/bash
set -e # this makes the whole script exit on any error.
#fill these variables with the arguments given to this script
sample=$1
fq1=$2
fq2=$3
logDir=MyLogDirectory # this is where all the files to keep track of progress will go.
mkdir -p MyLogDirectory # make the directory where log files will go, if it doesn't exist already echo running pipeline for $sample
if [ ! -e $logDir/$sample.fastqc.done ] #run this code only if $logDir/$sample.fastqc.done is missing
then
        echo Performing fastqc of sample $sample with the following fastqs:
        ls /projects/bmeg/A2/$fq1 /projects/bmeg/A2/$fq2

        #enter commands to run fastqc here
        fastqc $fq1
        fastqc $fq2

        touch $logDir/$sample.fastqc.done #create the file that we were looking for at the beginning of this if statement so that this same code is not run next time
else # $logDir/$sample.fastqc.done was not missing
echo Already performed fastqc of $sample
fi
# here is where you will be including the additional steps to take you from fastq.gz to sorted BAM containing only uniquely mapping reads.
# create paired-end alignment of the $sample sequences to the human genome hg38 using bowtie2
echo bowtie2 paired-end alignment of $sample sequences to human genome hg38
bowtie2 -x /projects/bmeg/indexes/hg38/hg38_bowtie2_index -1 $fq1 -2 $fq2 -S $sample.bowtie2.sam
# use samtools to convert sam file to bam
echo samtools to convert sam file to bam of $sample
samtools view -S -b -h $sample.bowtie2.sam > $sample.samtools.bam
echo unsorted samtools output
samtools view $sample.samtools.bam | cut -f1,3,4 | head -5
# sort bam file using sambamba
echo sort $sample bam using sambamba
sambamba sort $sample.samtools.bam
echo sambamba sorted $sample
samtools view $sample.samtools.sorted.bam | cut -f1,3,4 | head -5
echo wc of $sample sorted
samtools view $sample.samtools.sorted.bam | cut -f1,3,4 | wc
# Sort uniquely mapped reads only
sambamba view -h -f "bam" -F "[XS] == null and not unmapped and not duplicate" $sample.samtools.sorted.bam > $sample.sambamba.sorted.bam
echo view uniquely mapped reads for $sample
samtools view $sample.sambamba.sorted.bam | cut -f1,3,4 | head -5
echo wc of uniquely mapped reads for $sample
samtools view $sample.sambamba.sorted.bam | cut -f1,3,4 | wc
  
```
  
```{bash, eval=FALSE}

# Output of script
2614910 reads; of these:
  2614910 (100.00%) were paired; of these:
    220030 (8.41%) aligned concordantly 0 times
    2129747 (81.45%) aligned concordantly exactly 1 time
    265133 (10.14%) aligned concordantly >1 times
    ----
    220030 pairs aligned concordantly 0 times; of these:
      169387 (76.98%) aligned discordantly 1 time
    ----
    50643 pairs aligned 0 times concordantly or discordantly; of these:
      101286 mates make up the pairs; of these:
        34076 (33.64%) aligned 0 times
        30311 (29.93%) aligned exactly 1 time
        36899 (36.43%) aligned >1 times
99.35% overall alignment rate
samtools to convert sam file to bam of $sample
unsorted samtools output
SRR12694366.1000000     chr3    46631700
SRR12694366.1000000     chr3    46631857
SRR12694366.10000029    chr3    77719001
SRR12694366.10000029    chr3    77718984
SRR12694366.10000035    chr3    40015576
sort iPSC_input bam using sambamba
SRR12694366.7794671     chr12   133198776
SRR12694366.7794671     chr12   133198776
SRR12694366.26500791    chr3    9988
SRR12694366.33624963    chr3    10013
SRR12694366.33624963    chr3    10013
wc of iPSC_input sorted
4444490 13333470 156163784
view uniquely mapped reads for iPSC_input
SRR12694366.7794671     chr12   133198776
SRR12694366.26500791    chr3    9988
SRR12694366.33624963    chr3    10013
SRR12694366.33999302    chr3    10033
SRR12694366.4318510     chr3    10228
wc of uniquely mapped reads for iPSC_input
3859882 11579646 135643657
Already performed fastqc of H3K27me3
bowtie2 paired-end alignment of H3K27me3 sequences to human genome hg38
1891826 reads; of these:
  1891826 (100.00%) were paired; of these:
    171014 (9.04%) aligned concordantly 0 times
    1517253 (80.20%) aligned concordantly exactly 1 time
    203559 (10.76%) aligned concordantly >1 times
    ----
    171014 pairs aligned concordantly 0 times; of these:
      104276 (60.98%) aligned discordantly 1 time
    ----
    66738 pairs aligned 0 times concordantly or discordantly; of these:
      133476 mates make up the pairs; of these:
        54695 (40.98%) aligned 0 times
        49364 (36.98%) aligned exactly 1 time
        29417 (22.04%) aligned >1 times
98.55% overall alignment rate
samtools to convert sam file to bam of H3K27me3
unsorted samtools output
SRR12694360.1000001     chr3    156727688
SRR12694360.1000001     chr3    156727578
SRR12694360.10000059    chr3    59983210
SRR12694360.10000059    chr3    59983117
SRR12694360.1000008     chr3    8311099
sort H3K27me3 bam using sambamba
SRR12694360.31488484    chr12   48333021
SRR12694360.31488484    chr12   48333021
SRR12694360.386539      chr13   56447072
SRR12694360.386539      chr13   56447072
SRR12694360.9684526     chr3    9980
wc of H3K27me3 sorted
3783652 11350956 132947999
view uniquely mapped reads for H3K27me3
SRR12694360.31488484    chr12   48333021
SRR12694360.386539      chr13   56447072
SRR12694360.9684526     chr3    9980
SRR12694360.10316301    chr3    10148
SRR12694360.34275391    chr3    10226
wc of uniquely mapped reads for H3K27me3
3218587 9655761 113107772

# Runs as expected

```


# Authors and contributions

Following completion of your assignment, please fill out this section with the authors and their contributions to the '
assignment.  If you worked alone, only the author (e.g. your name and student ID) should be included.

Authors: Erik Lamoureux

Contributions: Assignment completed by Erik Lamoureux individually. 


